### PR TITLE
fix: RouterComponent replace methods to correctly handle previous/nextRoute

### PR DESCRIPTION
**Pull request by spydon**
# Description

Before when you only had two routes you'd get:
```
Bad state: No element

dart:core                                                  List.last
package:flame/src/components/router_component.dart 260:30  RouterComponent._popWithoutMinimumAssert
package:flame/src/components/router_component.dart 122:7   RouterComponent.pushReplacementNamed
test/components/router_component_test.dart 20:14           main.<fn>.<fn>
```

## Checklist

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [x] No, this PR is not a breaking change.

## Related Issues

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md

### CODE REVIEW

**Review by erickzanardo — APPROVED** [review]

**Review by munsterlander — COMMENTED** [review]
LGTM - odd that made it through, but it comes from pop with the same tests just modified.

Edit to clarify:

I just want to confirm that `pop()` does not have the same issue. 

```dart
void pop() {
    assert(
      _routeStack.length > 1,
      'Cannot pop the last route from the Router',
    );
    final route = _routeStack.removeLast();
    _adjustRoutesOrder();
    _adjustRoutesVisibility();
    route.didPop(_routeStack.last);
}
```

**Comment by spydon** [comment]
> I just want to confirm that `pop()` does not have the same issue.

It does indeed, good catch!
I'll write some tests that tests this.

**Comment by spydon** [comment]
Ah wait, this is wrong, `onPop` should return the next route, not the route that was popped.

**Comment by munsterlander** [comment]
> Ah wait, this is wrong, `onPop` should return the next route, not the route that was popped.

Ah! When I was adding this, I was confused why that was there, but not knowing the full router creation origin story, I was like well there had to be a reason.   The tests passed for what wa expected, but good to know.

**Comment by spydon** [comment]
> Ah! When I was adding this, I was confused why that was there, but not knowing the full router creation origin story, I was like well there had to be a reason. The tests passed for what was expected, but good to know.

Fixed it and added some extra tests that should cover this. :)

File: packages/flame/lib/src/components/router_component.dart [hunk 1]
```diff
@@ -255,9 +239,9 @@ class RouterComponent extends Component {
   }
 
   /// Local method to bypass [pop]'s assert
-  void _popWithoutMinimumAssert() {
+  void _replaceRoute(Route nextRoute) {
```
Comment (st-pasha) [inline packages/flame/lib/src/components/router_component.dart:242]: `_removeTopRoute`? This method doesn't actually replace anything.

File: packages/flame/test/components/router_component_test.dart [hunk 2]
```diff
@@ -4,6 +4,47 @@ import 'package:flame_test/flame_test.dart';
 import 'package:flutter/widgets.dart' hide Route, OverlayRoute;
 import 'package:flutter_test/flutter_test.dart';
 
+class _TestRoute extends Route {
```
Comment (st-pasha) [inline packages/flame/test/components/router_component_test.dart:7]: Could you please move this helper class to the bottom of the file? That way, when you open file you can immediately see `void main()` and can press the green arrow to start the tests.

**Review by st-pasha — APPROVED** [review]

File: packages/flame/lib/src/components/router_component.dart (hunk 1 above)
Comment (spydon) [reply packages/flame/lib/src/components/router_component.dart]: Done

File: packages/flame/test/components/router_component_test.dart (hunk 2 above)
Comment (spydon) [reply packages/flame/test/components/router_component_test.dart]: Done

**Pull request merged**